src/main/java/com/github/princesslana/smalld/Heartbeat.java

```diff
@@ -57,14 +64,38 @@ private void onHello(SmallD smalld, JsonObject d) {
     long interval = d.getInt("heartbeat_interval", -1);
 
     heartbeat =
-        heartbeatExecutor.scheduleAtFixedRate(
-            () -> sendHeartbeat(smalld), interval, interval, TimeUnit.MILLISECONDS);
+        heartbeatExecutor.schedule(
+            () -> runHeartbeatLoop(smalld, interval), interval, TimeUnit.MILLISECONDS);
+  }
+
+  private void runHeartbeatLoop(SmallD smalld, long interval) {
+    while (true) {
```

Comment:
It feels unnecessary to run our own loop here (if we're going to, we don't really need to use an Executor to do so).

We should be able to check we received the ack before sending a heartbeat?

Reply:
Yeah makes sense, the loop is unnecessary.